source/gateway.go

```diff
@@ -417,12 +420,23 @@ func (c *gatewayRouteResolver) hosts(rt gatewayRoute) ([]string, error) {
 		}
 		hostnames = append(hostnames, hosts...)
 	}
-	// This means that the route doesn't specify a hostname and should use any provided by
-	// attached Gateway Listeners. This is only useful for {HTTP,TLS}Routes, but it doesn't
-	// break {TCP,UDP}Routes.
-	if len(rt.Hostnames()) == 0 {
-		hostnames = append(hostnames, "")
+
+	hostNameAnnotation, hostNameAnnotationExists := rt.Metadata().Annotations[annotations.GatewayHostnameSourceKey]
+	if !hostNameAnnotationExists {
+		// This means that the route doesn't specify a hostname and should use any provided by
+		// attached Gateway Listeners. This is only useful for {HTTP,TLS}Routes, but it doesn't
+		// break {TCP,UDP}Routes.
+		if len(rt.Hostnames()) == 0 {
+			hostnames = append(hostnames, "")
+		}
+
+		hostnames = append(hostnames, annotationHostnames...)
 	}
+
+	if strings.ToLower(hostNameAnnotation) == gatewayHostnameSourceAnnotationOnlyValue {
+		return annotationHostnames, nil
+	}
+
 	return hostnames, nil
```

Comment:
You're not checking for `gatewayHostnameSourceDefinedHostsOnlyValue`, so any invalid value will return only the defined hostnames.
In this case, the ingress implementation returns an empty list.
It might be worth logging a warning too.

You also build hostnames from both sources and eventually discard them.
Wouldn’t it be better to only add hostnames when they’re actually needed?